internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,248 @@
+		return
+	}
+
+	s.baseContext, s.baseCancel = context.WithCancel(context.Background())
+	go s.start()
+	// TODO (lruch): start go routine to monitor and update progress of running jobs
+	// TODO (lruch): add watchdogs to monitor defunct jobs
+	s.started.Store(true)
+}
+
+func (s *Scheduler) start() {
+	s.logger.Debug("starting scheduling loop")
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-s.baseContext.Done():
+			s.logger.Debug("scheduling loop shutting down")
+			return
+
+		case <-timer.C:
+			s.logger.Debug("waking up to run jobs")
+
+			repo, err := s.jobRepoFn()
+			if err != nil {
+				s.logger.Error("error creating job repo", "error", err)
+				break
+			}
+
+			runs, err := repo.RunJobs(s.baseContext, s.serverId, jobRepo.WithRunJobsLimit(s.runJobsLimit))
+			if err != nil {
+				s.logger.Error("error getting jobs to run from repo", "error", err)
+				break
+			}
+
+			for _, r := range runs {
+				err := s.runJob(r)
+				if err != nil {
+					s.logger.Error("error starting job", "error", err)
+					if _, inner := repo.FailRun(s.baseContext, r.PrivateId); inner != nil {
+						s.logger.Error("error updating failed job run", "error", inner)
+					}
+				}
+			}
+		}
+
+		s.logger.Debug("scheduling loop going back to sleep")
+		timer.Reset(s.runJobsInterval)
+	}
+}
+
+func (s *Scheduler) runJob(r *jobRepo.Run) error {
+	jobId := JobId(r.JobId)
+	j, ok := s.registeredJobs.Load(jobId)
+	job := j.(Job)
+	s.logger.Debug(fmt.Sprintf("starting job run %q for job (%v: %v)", r.PrivateId, jobId, job.Name()))
+	if !ok {
+		return fmt.Errorf("job (%v: %v) not registered on scheduler", jobId, job.Name())
+	}
+
+	s.l.Lock()
+	if _, ok := s.runningJobs[jobId]; ok {
+		s.l.Unlock()
+		return fmt.Errorf("job (%v: %v) is already running", jobId, job.Name())
+	}
+	jobContext, jobCancel := context.WithCancel(context.Background())
+	s.runningJobs[jobId] = runningJob{runId: r.PrivateId, cancelCtx: jobCancel}
+	s.l.Unlock()
+
+	go func() {
+		runErr := job.Run(jobContext)
+		repo, err := s.jobRepoFn()
+		if err != nil {
+			// Failed to create repo needed update run.  Best option is to log error,
+			// delete running job and return.
+			s.logger.Error("error creating job repo after job run", "error", err)
+			s.l.Lock()
+			delete(s.runningJobs, jobId)
+			s.l.Unlock()
+			return
+		}
+
+		switch runErr {
+		case nil:
+			s.logger.Debug(fmt.Sprintf("job run %q for job (%v: %v) complete", r.PrivateId, jobId, job.Name()))
+			if _, inner := repo.CompleteRun(jobContext, r.PrivateId, job.NextRunIn()); inner != nil {
+				s.logger.Error("error updating completed job run", "error", inner)
+			}
+		default:
+			s.logger.Debug(fmt.Sprintf("job run %q for job (%v: %v) failed: %v", r.PrivateId, jobId, job.Name(), runErr))
+			if _, inner := repo.FailRun(jobContext, r.PrivateId); inner != nil {
+				s.logger.Error("error updating failed job run", "error", inner)
+			}
+		}
+
+		s.l.Lock()
+		delete(s.runningJobs, jobId)
```

Comment:
The cancel function for the job's context is never called which will leak resources.